File: python/open_guardrail/guards/academic_integrity_output.py
"""Prevent AI from completing academic assignments."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult

_ASSIGN_REQ = re.compile(
    r"\b(?:write\s+my\s+essay"
    r"|do\s+my\s+homework"
    r"|solve\s+this\s+exam"
    r"|answer\s+these\s+test\s+questions"
    r"|complete\s+this\s+assignment)\b",
    re.IGNORECASE,
)
_COMPLETION = re.compile(
    r"\b(?:here\s+is\s+your\s+completed"
    r"\s+assignment"
    r"|here\s+is\s+the\s+essay"
    r"|I've\s+written\s+the\s+paper"
    r"|//\s*Solution:"
    r"|def\s+solution\("
    r"|Answer:\s*[A-E]"
    r"|The\s+correct\s+answer\s+is)\b",
    re.IGNORECASE,
)
_EDU_FRAME = re.compile(
    r"\b(?:here'?s?\s+how\s+to\s+approach"
    r"|let\s+me\s+explain\s+the\s+concept"
    r"|step[- ]by[- ]step\s+guide"
    r"|learning\s+objective)\b",
    re.IGNORECASE,
)


class _AcademicIntegrityOutput:
    def __init__(
        self, *, action: str = "block"
    ) -> None:
        self.name = "academic-integrity-output"
        self.action = action

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()
        has_assign = bool(
            _ASSIGN_REQ.search(text)
        )
        has_complete = bool(
            _COMPLETION.search(text)
        )
        has_edu = bool(_EDU_FRAME.search(text))
        triggered = (
            has_assign or has_complete
        ) and not has_edu
        elapsed = (
            time.perf_counter() - start
        ) * 1000
        if not triggered:
            return GuardResult(
                guard_name=(
                    "academic-integrity-output"
                ),
                passed=True,
                action="allow",
                latency_ms=round(elapsed, 2),
            )
        return GuardResult(
            guard_name="academic-integrity-output",
            passed=False,
            action=self.action,
            message=(
                "Academic assignment completion"
                " detected without educational"
                " framing"
            ),
            latency_ms=round(elapsed, 2),
            details={
                "assignment_request": has_assign,
                "completion_detected": has_complete,
                "educational_framing": has_edu,
            },
        )


def academic_integrity_output(
    *, action: str = "block"
) -> _AcademicIntegrityOutput:
    return _AcademicIntegrityOutput(
        action=action
    )

File: packages/guards/src/academic-integrity-output.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';
interface AcademicIntegrityOutputOptions { action: 'block' | 'warn'; }
const ASSIGN_REQ = /\b(?:write\s+my\s+essay|do\s+my\s+homework|solve\s+this\s+exam|answer\s+these\s+test\s+questions|complete\s+this\s+assignment)\b/i;
const COMPLETION = /\b(?:here\s+is\s+your\s+completed\s+assignment|here\s+is\s+the\s+essay|I've\s+written\s+the\s+paper|\/\/\s*Solution:|def\s+solution\(|Answer:\s*[A-E]|The\s+correct\s+answer\s+is)\b/i;
const EDU_FRAME = /\b(?:here'?s?\s+how\s+to\s+approach|let\s+me\s+explain\s+the\s+concept|step[- ]by[- ]step\s+guide|learning\s+objective)\b/i;
export function academicIntegrityOutput(options: AcademicIntegrityOutputOptions): Guard {
  return { name: 'academic-integrity-output', version: '0.1.0', description: 'Prevent AI from completing academic assignments', category: 'content',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const hasAssignment = ASSIGN_REQ.test(text);
      const hasCompletion = COMPLETION.test(text);
      const hasEduFrame = EDU_FRAME.test(text);
      const triggered = (hasAssignment || hasCompletion) && !hasEduFrame;
      return { guardName: 'academic-integrity-output', passed: !triggered, action: triggered ? options.action : 'allow',
        message: triggered ? 'Academic assignment completion detected without educational framing' : undefined,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { assignmentRequest: hasAssignment, completionDetected: hasCompletion, educationalFraming: hasEduFrame } : undefined,
      };
    },
  };
}
